# DataFrameで結合などの処理を確認する

## 1. ライブラリのimport

In [1]:
import os

import numpy as np
import pandas as pd

## 2. データ読み込み

In [2]:
dir_path = "./data/sample_join"
file_name = "sample.csv"

file_path = os.path.join(dir_path, file_name)

In [3]:
header_names = ["category_id", "created_at", "data"]

In [4]:
# 元のCSVファイルにはヘッダーがついていないので、プログラム側でヘッダー名を指定する
df = pd.read_csv(file_path, header=None, names=header_names)

df

category_id                 created_at   data
0            1  2022-12-27T12:00:00+09:00  10.00
1            1  2022-12-27T12:00:01+09:00  10.01
2            2  2022-12-27T12:00:01+09:00  20.01
3            1  2022-12-27T12:00:03+09:00  10.03
4            1  2022-12-27T12:00:04+09:00  10.04
5            2  2022-12-27T12:00:00+09:00  20.00
6            2  2022-12-27T12:00:03+09:00  20.03
7            1  2022-12-27T12:00:02+09:00  10.02
8            2  2022-12-27T12:00:02+09:00  20.02
9            2  2022-12-27T12:00:04+09:00  20.04

## 3. 処理

読み込んだCSVデータに対して、次のようなCSVを出力することを考える。

| category_id | created_at | data | rank |
| --- | --- | --- | --- |
| 1	| 2022-12-27T12:00:00+09:00 | 10.00 | 0 |
| 1	| 2022-12-27T12:00:01+09:00 | 10.01 | 1 |
| 2	| 2022-12-27T12:00:01+09:00 | 20.01 | 1 |
| 1	| 2022-12-27T12:00:03+09:00 | 10.03 | 3 |
| 1	| 2022-12-27T12:00:04+09:00 | 10.04 | 4 |
| 2	| 2022-12-27T12:00:00+09:00 | 20.00 | 0 |
| 2	| 2022-12-27T12:00:03+09:00 | 20.03 | 3 |
| 1	| 2022-12-27T12:00:02+09:00 | 10.02 | 2 |
| 2	| 2022-12-27T12:00:02+09:00 | 20.02 | 2 |
| 2	| 2022-12-27T12:00:04+09:00 | 20.04 | 4 |


出力するCSVデータの取得/処理要件は次の通りである。

- 元のCSVデータと同じ行数、同じ順序である
- `rank`列を追加する
    - `rank`は同一の`category_id`グループにおいて、`created_at`を昇順ソートした順番である (0始まり)

### 参考

`rank`列を見やすく整形した表は以下の通りである。

| category_id | created_at | data | rank |
| --- | --- | --- | --- |
| 1 | 2022-12-27T12:00:00+09:00 | 10.00 | 0 |
| 1 | 2022-12-27T12:00:01+09:00 | 10.01 | 1 |
| 1 | 2022-12-27T12:00:02+09:00 | 10.02 | 2 |
| 1	| 2022-12-27T12:00:03+09:00 | 10.03 | 3 |
| 1	| 2022-12-27T12:00:04+09:00 | 10.04 | 4 |
| 2 | 2022-12-27T12:00:00+09:00 | 20.00 | 0 |
| 2 | 2022-12-27T12:00:01+09:00 | 20.01 | 1 |
| 2 | 2022-12-27T12:00:02+09:00 | 20.02 | 2 |
| 2 | 2022-12-27T12:00:03+09:00 | 20.03 | 3 |
| 2 | 2022-12-27T12:00:04+09:00 | 20.04 | 4 |